src/main/java/de/tum/in/www1/artemis/config/migration/entries/MigrationEntry20230808_203400.java

```diff
@@ -0,0 +1,460 @@
+            if (!finished) {
+                log.error(ERROR_MESSAGE);
+                throw new RuntimeException(ERROR_MESSAGE);
+            }
+        }
+        catch (InterruptedException e) {
+            log.error(ERROR_MESSAGE);
+            throw new RuntimeException(e);
+        }
+
+        evaluateErrorMap(errorMap);
+    }
+
+    /**
+     * Executes the migration for a batch of programming exercises. Expected to be in a separate thread.
+     *
+     * @param start         The start index of the batch page
+     * @param end           The end index of the batch page
+     * @param exerciseCount The total number of exercises for this thread
+     * @param errorMap      The error map to write errors to
+     */
+    private void executeBatchedMigration(long start, long end, long exerciseCount, Map<ProgrammingExercise, Boolean> errorMap) {
+        // do batches
+        for (int batchStartIndex = (int) start, batch = 0; batchStartIndex < end; batchStartIndex++, batch = batch + BATCH_SIZE) {
+            var batchStart = System.currentTimeMillis();
+            Pageable pageable = PageRequest.of(batchStartIndex, BATCH_SIZE);
+
+            // TODO: in the unlikely case the buildPlanId is null, we could skip the template participation
+            var templateParticipationPage = templateProgrammingExerciseParticipationRepository.findAll(pageable);
+            log.info("Found {} template programming exercises to migrate in batch", templateParticipationPage.getTotalElements());
+            templateParticipationPage.map(templateParticipation -> {
+                if (!ciMigrationService.orElseThrow().buildPlanExists(templateParticipation.getProgrammingExercise().getProjectKey(),
+                        templateParticipation.getProgrammingExercise().getTemplateBuildPlanId())) {
+                    log.warn("Build plan {} does not exist for exercise {}", templateParticipation.getProgrammingExercise().getTemplateBuildPlanId(),
+                            templateParticipation.getProgrammingExercise().getId());
+                    log.info("Skipping migration for template of exercise {} and setting build_plan_id to null", templateParticipation.getProgrammingExercise().getId());
+                    templateParticipation.setBuildPlanId(null);
+                    templateProgrammingExerciseParticipationRepository.save(templateParticipation);
+                    return null;
+                }
+                log.info("Migrating build plan with name {} for exercise {}",
+                        templateParticipation.getProgrammingExercise().getProjectKey() + "-" + templateParticipation.getBuildPlanId(),
+                        templateParticipation.getProgrammingExercise().getId());
+                return templateParticipation;
+            }).filter(Objects::nonNull).forEach(templateParticipation -> {
+                var startMs = System.currentTimeMillis();
```

Comment:
I think @Strohgelaender told me about some better API to use once... Can you ask him about that? I forgot 😅 It was something like "performance nanoseconds" or so.

Reply:
It's just a nice addition, we don't need actual accuracy.